ir/expression_ref.h

```diff
@@ -0,0 +1,82 @@
+#ifndef IR_EXPRESSION_REF_H_
+#define IR_EXPRESSION_REF_H_
+
+#include "ir.h"
+
+namespace IR {
+
+class ERef {
+    const Expression *expr;
+
+    template<class T> class ERef_specific {
+        const T *expr;
+        friend class ERef;
+
+     public:
+        ERef_specific(const T *e) : expr(e) {}  // NOLINT(runtime/explicit)
+        const T &operator*() const { return *expr; }
+        const T *operator->() const { return expr; }
+        operator const T *() const { return expr; }
+    };
+
+ public:
+    ERef(const Expression *e) : expr(e) {}   // NOLINT(runtime/explicit)
+    ERef(const Expression &e) : expr(&e) {}   // NOLINT(runtime/explicit)
```

Comment:
Are Expression objects always allocated via new? If you pass a temporary here you'll be storing a pointer to the stack. That's a pretty big footgun. Presumably if they *are* always allocated via new, the constructor that takes a pointer is sufficient, so I'd suggest only providing that one or else performing an actual copy here.